IdentityCore/src/cache/ios/MSIDKeychainTokenCache.m

```diff
@@ -22,30 +22,186 @@
+    [query setObject:key.service forKey:(id)kSecAttrService];
+    [query setObject:key.account forKey:(id)kSecAttrAccount];
+    [query setObject:self.accessGroup forKey:(id)kSecAttrAccessGroup];
+    // Backward compatibility with ADAL.
+    [query setObject:[s_libraryString dataUsingEncoding:NSUTF8StringEncoding] forKey:(id)kSecAttrGeneric];
+    // Backward compatibility with MSAL.
+    [query setObject:[NSNumber numberWithUnsignedInt:item.tokenType] forKey:(id)kSecAttrType];
+    
+    MSID_LOG_INFO(context, @"Trying to update keychain item...");
+    OSStatus status = SecItemUpdate((CFDictionaryRef)query, (CFDictionaryRef)@{(id)kSecValueData : itemData});
+    MSID_LOG_INFO(context, @"Keychain update status: %d", status);
+    if (status == errSecItemNotFound)
+    {
+        [query setObject:itemData forKey:(id)kSecValueData];
+        [query setObject:(id)kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly forKey:(id)kSecAttrAccessible];
+        
+        MSID_LOG_INFO(context, @"Trying to add keychain item...");
+        status = SecItemAdd((CFDictionaryRef)query, NULL);
+        MSID_LOG_INFO(context, @"Keychain add status: %d", status);
+    }
+    
+    if (status != errSecSuccess)
+    {
+        if (error)
+        {
+            *error = [[NSError alloc] initWithDomain:@"MSIDErrorDomain" code:status userInfo:@{NSLocalizedDescriptionKey:@"Failed to set item into keychain."}];
+        }
+        MSID_LOG_ERROR(context, @"Failed to set item into keychain (status: %d)", status);
+    }
+    
+    return status == errSecSuccess;
 }
 
 - (MSIDToken *)itemWithKey:(MSIDTokenCacheKey *)key
                 serializer:(id<MSIDTokenSerializer>)serializer
                    context:(id<MSIDRequestContext>)context
                      error:(NSError **)error
 {
-    return nil;
+    MSID_LOG_INFO(context, @"itemWithKey:serializer:context:error:");
+    return [self itemsWithKey:key serializer:serializer context:context error:error].firstObject;
 }
 
 - (BOOL)removeItemWithKey:(MSIDTokenCacheKey *)key
                   context:(id<MSIDRequestContext>)context
                     error:(NSError **)error
 {
+    assert(key);
+    
+    MSID_LOG_INFO(context, @"Remove keychain item, key info (account: %@ service: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service));
+    MSID_LOG_INFO_PII(context, @"Remove keychain item, key info (account: %@ service: %@)", key.account, key.service);
+    
+    if (!key.service || !key.account)
+    {
+        if (error)
+        {
+            *error = [[NSError alloc] initWithDomain:@"MSIDErrorDomain" code:-1 userInfo:@{NSLocalizedDescriptionKey:@"Key is not valid. Make sure service and account are not nil."}];
+        }
+        MSID_LOG_ERROR(context, @"Remove keychain item with invalid key.");
+        return NO;
+    }
+    
+    NSMutableDictionary *query = [@{(id)kSecClass : (id)kSecClassGenericPassword} mutableCopy];
```

Comment:
I think this query should also specify access group?

Reply:
fixed